File: north/cli/commands/mcp.py
"""`north mcp start|stop|status|run` — manage the on-demand MCP server.

The server is a uvicorn process running ``north.service.main:app``. It is
user-managed (no systemd): ``start`` spawns it detached, ``stop`` signals it,
``status`` checks it, and ``run`` runs it in the foreground. The board is passed
to the server via the ``NORTH_BOARD`` env var; the port comes from
``north/config.yml``.
"""

import argparse
import os
import signal
import subprocess
import sys
from pathlib import Path

from north.cli.errors import CLIError
from north.core.board import load_config, locate_board

_NORTH_HOME = Path.home() / ".north"
_PID_FILE = _NORTH_HOME / "mcp.pid"
_LOG_FILE = _NORTH_HOME / "mcp.log"
_HOST = "127.0.0.1"


def _read_pid() -> int | None:
    if not _PID_FILE.exists():
        return None
    try:
        pid = int(_PID_FILE.read_text().strip())
    except ValueError:
        return None
    return pid if _alive(pid) else None


def _alive(pid: int) -> bool:
    try:
        os.kill(pid, 0)
    except OSError:
        return False
    return True


def start(args: argparse.Namespace) -> int:
    board = locate_board()
    port = load_config(board).mcp_port
    if (pid := _read_pid()) is not None:
        raise CLIError(f"MCP server already running (pid {pid}) — `north mcp stop` first")
    _NORTH_HOME.mkdir(parents=True, exist_ok=True)
    env = {**os.environ, "NORTH_BOARD": str(board)}
    log = _LOG_FILE.open("a", encoding="utf-8")
    proc = subprocess.Popen(
        [sys.executable, "-m", "uvicorn", "north.service.main:app",
         "--host", _HOST, "--port", str(port)],
        stdout=log,
        stderr=log,
        stdin=subprocess.DEVNULL,
        start_new_session=True,
        env=env,
    )
    _PID_FILE.write_text(str(proc.pid))
    print(f"MCP server started (pid {proc.pid}) at http://{_HOST}:{port}/mcp")
    print(f"logs: {_LOG_FILE}")
    return 0


def stop(args: argparse.Namespace) -> int:
    pid = _read_pid()
    if pid is None:
        _PID_FILE.unlink(missing_ok=True)
        print("MCP server is not running.")
        return 0
    os.kill(pid, signal.SIGTERM)
    _PID_FILE.unlink(missing_ok=True)
    print(f"Stopped MCP server (pid {pid}).")
    return 0


def status(args: argparse.Namespace) -> int:
    pid = _read_pid()
    if pid is None:
        print("MCP server: stopped")
        return 0
    port = load_config(locate_board()).mcp_port
    print(f"MCP server: running (pid {pid}) at http://{_HOST}:{port}/mcp")
    return 0


def run(args: argparse.Namespace) -> int:
    import uvicorn

    board = locate_board()
    os.environ["NORTH_BOARD"] = str(board)
    uvicorn.run("north.service.main:app", host=_HOST, port=load_config(board).mcp_port)
    return 0

File: internal/cli/mcp.go
package cli

import (
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strconv"
	"strings"
	"syscall"

	"github.com/SamP-S/north/internal/board"
	nerrors "github.com/SamP-S/north/internal/errors"
	"github.com/SamP-S/north/internal/service"
	"github.com/spf13/cobra"
)

const mcpHost = "127.0.0.1"

func northHome() string { return filepath.Join(mustHome(), ".north") }
func pidFile() string   { return filepath.Join(northHome(), "mcp.pid") }
func logFile() string   { return filepath.Join(northHome(), "mcp.log") }

func mustHome() string {
	home, err := os.UserHomeDir()
	if err != nil {
		return os.TempDir()
	}
	return home
}

func newMCPCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "mcp",
		Short: "manage the on-demand MCP server",
	}
	cmd.AddCommand(
		&cobra.Command{Use: "start", Short: "start the MCP server (detached)", Args: cobra.NoArgs, RunE: mcpStart},
		&cobra.Command{Use: "stop", Short: "stop the MCP server", Args: cobra.NoArgs, RunE: mcpStop},
		&cobra.Command{Use: "status", Short: "show MCP server status", Args: cobra.NoArgs, RunE: mcpStatus},
		&cobra.Command{Use: "run", Short: "run the MCP server in the foreground", Args: cobra.NoArgs, RunE: mcpRun},
	)
	return cmd
}

// readPID returns the running server's PID, or 0 if not running.
func readPID() int {
	data, err := os.ReadFile(pidFile())
	if err != nil {
		return 0
	}
	pid, err := strconv.Atoi(strings.TrimSpace(string(data)))
	if err != nil || !alive(pid) {
		return 0
	}
	return pid
}

func alive(pid int) bool {
	proc, err := os.FindProcess(pid)
	if err != nil {
		return false
	}
	return proc.Signal(syscall.Signal(0)) == nil
}

func mcpStart(cmd *cobra.Command, _ []string) error {
	boardDir, err := board.LocateBoard("")
	if err != nil {
		return err
	}
	cfg, err := board.LoadConfig(boardDir)
	if err != nil {
		return err
	}
	if pid := readPID(); pid != 0 {
		return nerrors.Conflict(fmt.Sprintf("MCP server already running (pid %d) — `north mcp stop` first", pid))
	}
	if err := os.MkdirAll(northHome(), 0o755); err != nil {
		return err
	}
	exe, err := os.Executable()
	if err != nil {
		return err
	}
	log, err := os.OpenFile(logFile(), os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0o644)
	if err != nil {
		return err
	}
	defer log.Close()
	proc := exec.Command(exe, "mcp", "run")
	proc.Env = append(os.Environ(), "NORTH_BOARD="+boardDir)
	proc.Stdout = log
	proc.Stderr = log
	proc.SysProcAttr = &syscall.SysProcAttr{Setsid: true}
	if err := proc.Start(); err != nil {
		return err
	}
	pid := proc.Process.Pid
	if err := os.WriteFile(pidFile(), []byte(strconv.Itoa(pid)), 0o644); err != nil {
		return err
	}
	_ = proc.Process.Release()
	cmd.Printf("MCP server started (pid %d) at http://%s:%d/mcp\n", pid, mcpHost, cfg.MCPPort)
	cmd.Printf("logs: %s\n", logFile())
	return nil
}

func mcpStop(cmd *cobra.Command, _ []string) error {
	pid := readPID()
	if pid == 0 {
		_ = os.Remove(pidFile())
		cmd.Println("MCP server is not running.")
		return nil
	}
	proc, err := os.FindProcess(pid)
	if err != nil {
		return err
	}
	_ = proc.Signal(syscall.SIGTERM)
	_ = os.Remove(pidFile())
	cmd.Printf("Stopped MCP server (pid %d).\n", pid)
	return nil
}

func mcpStatus(cmd *cobra.Command, _ []string) error {
	pid := readPID()
	if pid == 0 {
		cmd.Println("MCP server: stopped")
		return nil
	}
	boardDir, err := board.LocateBoard("")
	if err != nil {
		return err
	}
	cfg, err := board.LoadConfig(boardDir)
	if err != nil {
		return err
	}
	cmd.Printf("MCP server: running (pid %d) at http://%s:%d/mcp\n", pid, mcpHost, cfg.MCPPort)
	return nil
}

func mcpRun(cmd *cobra.Command, _ []string) error {
	boardDir, err := board.LocateBoard("")
	if err != nil {
		return err
	}
	if os.Getenv("NORTH_BOARD") == "" {
		_ = os.Setenv("NORTH_BOARD", boardDir)
	}
	cfg, err := board.LoadConfig(boardDir)
	if err != nil {
		return err
	}
	addr := fmt.Sprintf("%s:%d", mcpHost, cfg.MCPPort)
	return service.Serve(addr, service.LoadToken())
}
